# Data inspection and Wrangling

## Objective

Understand what data is available and what are the main issues.

After inspecting the data I perform the subsetting to smaller datasets containing only the data of interest.

The data cleaning is done on the wrangled subsets.

The cleaned data will be shortly characterized.

Afterwards the cleaned data is stored as csv. - since the files are not too big, also to provide compatibility with Excel.


## Summary

#### 1. Import libraries and dataframes
- Renaming of the datasets

#### 2. Data Inspection
- 2.1. What I am checking
- 2.2. Check df_co2
- 2.3. Check df_hdi
- 2.4. Check df_forest
- 2.5. Check df_deforest
- 2.6. Check df_energy
- 2.7. Check df_elect

#### 3. Data Wrangling
- 3.1. What are the constraints
- 3.2. Subsetting df_co2
- 3.3. Subsetting df_hdi
- 3.4. Subsetting df_forest
- 3.5. Subsetting df_deforest
- 3.6. Subsetting df_energy
- 3.7. Subsetting df_elect

#### 4. Data Cleaning
- 4.1. What I am cleaning
- 4.2. Cleaning df_co2
- 4.3. Cleaning df_hdi
- 4.4. Cleaning df_forest
- 4.5. Cleaning df_deforest
- 4.6. Cleaning df_energy
- 4.7. Cleaning df_elect 

#### 5. Data Understanding
- 5.1. What I am checking
- 5.2. Key info df_co2
- 5.3. Cleaning df_hdi
- 5.4. Cleaning df_forest
- 5.5. Cleaning df_deforest
- 5.6. Cleaning df_energy
- 5.7. Cleaning df_elect 

#### 6. Exporting dataframes


<hr style="border: none; height: 3px; background-color: #333;" />

## 1. Import libraries and dataframes 

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Importing the dataframes

# creates a path with the source folder

path = r'C:\Users\Marc\Mariana_CareerFoundry\Achiev6'

# importing the dataframes

df_co2 = pd.read_csv(os.path.join(path,'02_Data','2-1_Original', 'Owid', 'owid-co2-data.csv'), index_col = False)

df_hdi = pd.read_csv(os.path.join(path,'02_Data','2-1_Original','Owid', 'inequality-adjusted-human-development-index.csv'), index_col = False)

df_forest = pd.read_csv(os.path.join(path,'02_Data','2-1_Original','Owid', 'forest-area-as-share-of-land-area.csv'), index_col = False)

df_deforest = pd.read_csv(os.path.join(path,'02_Data','2-1_Original','Owid', 'annual-deforestation.csv'), index_col = False)

df_energy = pd.read_csv(os.path.join(path,'02_Data','2-1_Original','Owid', 'renewable-share-energy.csv'), index_col = False)

df_elect = pd.read_csv(os.path.join(path,'02_Data','2-1_Original', 'Owid', 'share-electricity-renewables.csv'), index_col = False)

The datasets were renamed to represent:

#### df_co2 = CO2 emissions (and many other indices)

#### df_hdi = (Inequality adjusted) human development index

#### df_forest = Forest area as share of land area

#### df_defor = Annual deforestation

#### df_energy = Share of renewable sources on energy matrix

#### df_elect = Share of renewable sources on electricity generation

<hr style="border: none; height: 3px; background-color: #333;" />

## 2. Data Inspection

### 2.1.What I am checking

For each dataset I will check:
- headline
- shape
- info
- basic statistics (describe)
- value counts of the "country" entries
- value counts of the "year" entries

### 2.2. Checking df_co2

In [3]:
df_co2.head()

country  year iso_code  population  gdp  cement_co2  \
0  Afghanistan  1850      AFG   3752993.0  NaN         NaN   
1  Afghanistan  1851      AFG   3767956.0  NaN         NaN   
2  Afghanistan  1852      AFG   3783940.0  NaN         NaN   
3  Afghanistan  1853      AFG   3800954.0  NaN         NaN   
4  Afghanistan  1854      AFG   3818038.0  NaN         NaN   

   cement_co2_per_capita  co2  co2_growth_abs  co2_growth_prct  ...  \
0                    NaN  NaN             NaN              NaN  ...   
1                    NaN  NaN             NaN              NaN  ...   
2                    NaN  NaN             NaN              NaN  ...   
3                    NaN  NaN             NaN              NaN  ...   
4                    NaN  NaN             NaN              NaN  ...   

   share_global_other_co2  share_of_temperature_change_from_ghg  \
0                     NaN                                   NaN   
1                     NaN                                 0.157   
2                     NaN                                 0.156   
3                     NaN                                 0.156   
4                     NaN                                 0.155   

   temperature_change_from_ch4  temperature_change_from_co2  \
0                          NaN                          NaN   
1                          0.0                          0.0   
2                          0.0                          0.0   
3                          0.0                          0.0   
4                          0.0                          0.0   

   temperature_change_from_ghg  temperature_change_from_n2o  total_ghg  \
0                          NaN                          NaN        NaN   
1                          0.0                          0.0        NaN   
2                          0.0                          0.0        NaN   
3                          0.0                          0.0        NaN   
4                          0.0                          0.0        NaN   

   total_ghg_excluding_lucf  trade_co2  trade_co2_share  
0                       NaN        NaN              NaN  
1                       NaN        NaN              NaN  
2                       NaN        NaN              NaN  
3                       NaN        NaN              NaN  
4                       NaN        NaN              NaN  

[5 rows x 79 columns]

In [4]:
df_co2.shape

(47415, 79)

In [5]:
df_co2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47415 entries, 0 to 47414
Data columns (total 79 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   country                                    47415 non-null  object 
 1   year                                       47415 non-null  int64  
 2   iso_code                                   39548 non-null  object 
 3   population                                 39414 non-null  float64
 4   gdp                                        15211 non-null  float64
 5   cement_co2                                 23764 non-null  float64
 6   cement_co2_per_capita                      22017 non-null  float64
 7   co2                                        30308 non-null  float64
 8   co2_growth_abs                             28157 non-null  float64
 9   co2_growth_prct                            24684 non-null  float64
 10  co2_including_luc     

In [6]:
df_co2.describe()

year    population           gdp    cement_co2  \
count  47415.000000  3.941400e+04  1.521100e+04  23764.000000   
mean    1926.781609  6.095120e+07  3.309008e+11      9.109400   
std       59.561600  3.289137e+08  3.090394e+12     66.662221   
min     1750.000000  2.220000e+02  4.998000e+07      0.000000   
25%     1883.000000  3.448688e+05  7.898213e+09      0.000000   
50%     1930.000000  2.449992e+06  2.760856e+10      0.029000   
75%     1976.000000  1.011714e+07  1.220233e+11      0.834000   
max     2022.000000  7.975105e+09  1.301126e+14   1692.404000   

       cement_co2_per_capita           co2  co2_growth_abs  co2_growth_prct  \
count           22017.000000  30308.000000    28157.000000     24684.000000   
mean                0.066766    391.272161        5.868311        19.097573   
std                 0.125316   1857.972724       59.929873       689.790696   
min                 0.000000      0.000000    -2032.366000      -100.000000   
25%                 0.000000      0.183000        0.000000        -0.847750   
50%                 0.008000      3.856000        0.026000         3.876000   
75%                 0.093000     47.257000        0.872000        10.699250   
max                 2.574000  37149.785000     1813.064000    102318.508000   

       co2_including_luc  co2_including_luc_growth_abs  ...  \
count       23320.000000                  23030.000000  ...   
mean          534.225707                      7.445586  ...   
std          2176.346564                     98.354360  ...   
min          -187.847000                  -2334.695000  ...   
25%             5.878750                     -0.739750  ...   
50%            27.679500                      0.071000  ...   
75%           124.716000                      2.561000  ...   
max         41637.617000                   2011.781000  ...   

       share_global_other_co2  share_of_temperature_change_from_ghg  \
count             2593.000000                          41280.000000   
mean                19.199870                              2.236230   
std                 33.432368                              9.115378   
min                  0.000000                             -0.782000   
25%                  0.249000                              0.003000   
50%                  1.431000                              0.072000   
75%                 19.405000                              0.337000   
max                100.000000                            100.000000   

       temperature_change_from_ch4  temperature_change_from_co2  \
count                 37840.000000                 41280.000000   
mean                      0.002991                     0.007369   
std                       0.016277                     0.041778   
min                      -0.001000                     0.000000   
25%                       0.000000                     0.000000   
50%                       0.000000                     0.000000   
75%                       0.001000                     0.001000   
max                       0.417000                     1.136000   

       temperature_change_from_ghg  temperature_change_from_n2o     total_ghg  \
count                 41280.000000                 37840.000000   6354.000000   
mean                      0.010630                     0.000490    790.430981   
std                       0.059565                     0.002922   3610.534250   
min                      -0.001000                     0.000000   -186.660000   
25%                       0.000000                     0.000000      8.450000   
50%                       0.000000                     0.000000     38.285000   
75%                       0.001000                     0.000000    153.595000   
max                       1.635000                     0.083000  49880.602000   

       total_ghg_excluding_lucf    trade_co2  trade_co2_share  
count               6354.000000  4398.000000      4397.000000  
mean                 759.384745    -7.157874   

In [7]:
df_co2.value_counts('country')

country
Cote d'Ivoire                     273
High-income countries             273
Bhutan                            273
Mauritania                        273
Europe                            273
                                 ... 
Panama Canal Zone (GCP)            30
St. Kitts-Nevis-Anguilla (GCP)     24
Ryukyu Islands (GCP)               21
Leeward Islands (GCP)               7
Kuwaiti Oil Fires (GCP)             1
Name: count, Length: 261, dtype: int64

In [8]:
df_co2.value_counts('year')

year
1955    258
1961    258
1972    258
1971    258
1970    258
       ... 
1774     29
1773     29
1772     29
1771     29
1750     29
Name: count, Length: 273, dtype: int64

### Comment

The meaning of all columns is provided by the source here:
https://github.com/owid/co2-data/blob/master/owid-co2-codebook.csv

This data set has way more data than necessary, first I will narrow down to the time of 1980 to 2022.

Then I will keep first the columns

 0   country                                    47415 non-null  object 
 
 1   year                                       47415 non-null  int64  
 
 2   iso_code                                   39548 non-null  object 
 
 3   population                                 39414 non-null  float64
 
 4   gdp                                        15211 non-null  float64
 
 7   co2                                        30308 non-null  float64
 
 9   co2_growth_prct                            24684 non-null  float64
 
 10  co2_including_luc                          23320 non-null  float64
 
 16  co2_per_capita                             26600 non-null  float64

 18  co2_per_unit_energy                        10240 non-null  float64

 20  coal_co2_per_capita                        24389 non-null  float64

 33  energy_per_capita                          10018 non-null  float64

 40  ghg_per_capita                             6354 non-null   float64
 
 41  land_use_change_co2                        37022 non-null  float64
 
 42  land_use_change_co2_per_capita             36313 non-null  float64
 
 47  oil_co2                                    25111 non-null  float64
 
 48  oil_co2_per_capita                         24380 non-null  float64

 51  primary_energy_consumption                 10060 non-null  float64

 53  share_global_co2                           28495 non-null  float64
 
 54  share_global_co2_including_luc             23320 non-null  float64
 
 57  share_global_cumulative_co2                28495 non-null  float64
 
 58  share_global_cumulative_co2_including_luc  23320 non-null  float64

 70  share_of_temperature_change_from_ghg       41280 non-null  float64
 
 72  temperature_change_from_co2                41280 non-null  float64
 
 73  temperature_change_from_ghg                41280 non-null  float64
 
 75  total_ghg                                  6354 non-null   float64
 
 76  total_ghg_excluding_lucf                   6354 non-null   float64
 
 77  trade_co2                                  4398 non-null   float64
 
 78  trade_co2_share                            4397 non-null   float64


### 2.3. Checking HDI

In [9]:
df_hdi.head()

Entity Code  Year  Inequality-adjusted Human Development Index
0  Afghanistan  AFG  2010                                        0.287
1  Afghanistan  AFG  2011                                        0.293
2  Afghanistan  AFG  2012                                        0.302
3  Afghanistan  AFG  2013                                        0.309
4  Afghanistan  AFG  2014                                        0.313

In [10]:
df_hdi.shape

(2106, 4)

In [11]:
df_hdi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2106 entries, 0 to 2105
Data columns (total 4 columns):
 #   Column                                       Non-Null Count  Dtype  
---  ------                                       --------------  -----  
 0   Entity                                       2106 non-null   object 
 1   Code                                         1976 non-null   object 
 2   Year                                         2106 non-null   int64  
 3   Inequality-adjusted Human Development Index  2106 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 65.9+ KB


In [12]:
df_hdi.describe()

Year  Inequality-adjusted Human Development Index
count  2106.000000                                  2106.000000
mean   2016.308642                                     0.571973
std       3.671640                                     0.197505
min    2010.000000                                     0.184000
25%    2013.000000                                     0.383000
50%    2016.000000                                     0.588000
75%    2019.750000                                     0.744000
max    2022.000000                                     0.910000

In [13]:
df_hdi.value_counts('Entity')

Entity
Afghanistan         13
Iceland             13
Malaysia            13
Maldives            13
Malta               13
                    ..
Samoa                4
Tuvalu               4
Tonga                4
Marshall Islands     3
Fiji                 2
Name: count, Length: 178, dtype: int64

In [14]:
df_hdi.value_counts('Year')

Year
2021    176
2022    176
2020    175
2019    174
2018    169
2016    168
2017    168
2015    165
2014    161
2013    157
2012    152
2011    142
2010    123
Name: count, dtype: int64

### Comment

This dataset has a relatively short time span when compared to the other ones

### 2.4. Checking df_forest

In [15]:
df_forest.head()

Entity Code  Year  Forest cover
0  Afghanistan  AFG  1990      1.850994
1  Afghanistan  AFG  1991      1.850994
2  Afghanistan  AFG  1992      1.850994
3  Afghanistan  AFG  1993      1.850994
4  Afghanistan  AFG  1994      1.850994

In [16]:
df_forest.shape

(7974, 4)

In [17]:
df_forest.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7974 entries, 0 to 7973
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Entity        7974 non-null   object 
 1   Code          6893 non-null   object 
 2   Year          7974 non-null   int64  
 3   Forest cover  7974 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 249.3+ KB


In [18]:
df_forest.describe()

Year  Forest cover
count  7974.000000   7974.000000
mean   2002.437171     32.785091
std      38.486417     24.470957
min    1000.000000      0.000000
25%    1997.000000     11.561380
50%    2005.000000     31.152338
75%    2013.000000     49.767952
max    2020.000000    242.742857

In [19]:
df_forest.value_counts('Entity')

Entity
France                              57
China                               48
Japan                               41
United States                       38
Philippines                         37
                                    ..
Sint Maarten (French part)          10
Bonaire, Sint Eustatius and Saba    10
Sudan                                9
South Sudan                          9
Taiwan                               9
Name: count, Length: 263, dtype: int64

### 2.5. Checking df_elect

In [20]:
df_deforest.head()

Entity Code  Year  Deforestation
0   Africa  NaN  1990      4096000.0
1   Africa  NaN  2000      4314000.0
2   Africa  NaN  2010      4444000.0
3   Africa  NaN  2015      4414000.0
4  Algeria  DZA  1990        14800.0

In [21]:
df_deforest.shape

(495, 4)

In [22]:
df_deforest.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Entity         495 non-null    object 
 1   Code           471 non-null    object 
 2   Year           495 non-null    int64  
 3   Deforestation  495 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 15.6+ KB


In [23]:
df_deforest.describe()

Year  Deforestation
count   495.000000   4.950000e+02
mean   2004.303030   2.916967e+05
std       9.371391   1.391850e+06
min    1990.000000   0.000000e+00
25%    2000.000000   1.300000e+02
50%    2010.000000   5.780000e+03
75%    2015.000000   5.109000e+04
max    2015.000000   1.581800e+07

In [24]:
df_forest.value_counts('Entity')

Entity
France                              57
China                               48
Japan                               41
United States                       38
Philippines                         37
                                    ..
Sint Maarten (French part)          10
Bonaire, Sint Eustatius and Saba    10
Sudan                                9
South Sudan                          9
Taiwan                               9
Name: count, Length: 263, dtype: int64

### 2.6. Checking df_energy

In [25]:
df_energy.head(25)

Entity Code  Year  Renewables (% equivalent primary energy)
0   Africa  NaN  1965                                  5.740281
1   Africa  NaN  1966                                  6.113969
2   Africa  NaN  1967                                  6.316580
3   Africa  NaN  1968                                  6.994845
4   Africa  NaN  1969                                  7.943916
5   Africa  NaN  1970                                  9.148895
6   Africa  NaN  1971                                  8.124994
7   Africa  NaN  1972                                  8.747272
8   Africa  NaN  1973                                  8.541620
9   Africa  NaN  1974                                  9.252387
10  Africa  NaN  1975                                  9.228769
11  Africa  NaN  1976                                  9.416380
12  Africa  NaN  1977                                  9.803094
13  Africa  NaN  1978                                  9.518462
14  Africa  NaN  1979                                  9.349527
15  Africa  NaN  1980                                  8.481442
16  Africa  NaN  1981                                  7.852752
17  Africa  NaN  1982                                  7.373427
18  Africa  NaN  1983                                  6.456112
19  Africa  NaN  1984                                  6.158577
20  Africa  NaN  1985                                  6.579342
21  Africa  NaN  1986                                  6.715967
22  Africa  NaN  1987                                  6.243963
23  Africa  NaN  1988                                  6.362116
24  Africa  NaN  1989                                  6.665460

In [26]:
df_energy.shape

(4879, 4)

In [27]:
df_energy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4879 entries, 0 to 4878
Data columns (total 4 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Entity                                    4879 non-null   object 
 1   Code                                      3529 non-null   object 
 2   Year                                      4879 non-null   int64  
 3   Renewables (% equivalent primary energy)  4879 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 152.6+ KB


In [28]:
df_energy.describe()

Year  Renewables (% equivalent primary energy)
count  4879.000000                               4879.000000
mean   1996.163558                                 11.309559
std      16.453654                                 11.893607
min    1965.000000                                  0.000000
25%    1983.000000                                  3.451529
50%    1997.000000                                  7.095672
75%    2010.000000                                 15.132521
max    2023.000000                                 74.301690

In [29]:
df_energy.value_counts('Entity')

Entity
Africa           59
Mexico           59
North America    59
Non-OECD (EI)    59
Netherlands      59
                 ..
Singapore         8
Oman              6
Iraq              3
Sri Lanka         1
Iceland           1
Name: count, Length: 100, dtype: int64

### 2.7. Checking df_elect

In [30]:
df_elect.head(30)

Entity Code  Year  Renewables - % electricity
0   ASEAN (Ember)  NaN  2000                   19.347086
1   ASEAN (Ember)  NaN  2001                   19.066320
2   ASEAN (Ember)  NaN  2002                   17.664303
3   ASEAN (Ember)  NaN  2003                   16.672487
4   ASEAN (Ember)  NaN  2004                   15.700016
5   ASEAN (Ember)  NaN  2005                   15.152674
6   ASEAN (Ember)  NaN  2006                   15.766005
7   ASEAN (Ember)  NaN  2007                   15.688036
8   ASEAN (Ember)  NaN  2008                   16.497475
9   ASEAN (Ember)  NaN  2009                   16.700829
10  ASEAN (Ember)  NaN  2010                   16.487564
11  ASEAN (Ember)  NaN  2011                   18.787397
12  ASEAN (Ember)  NaN  2012                   19.624712
13  ASEAN (Ember)  NaN  2013                   20.348326
14  ASEAN (Ember)  NaN  2014                   19.964802
15  ASEAN (Ember)  NaN  2015                   18.573849
16  ASEAN (Ember)  NaN  2016                   20.176855
17  ASEAN (Ember)  NaN  2017                   23.653614
18  ASEAN (Ember)  NaN  2018                   23.714792
19  ASEAN (Ember)  NaN  2019                   21.183790
20  ASEAN (Ember)  NaN  2020                   23.560490
21  ASEAN (Ember)  NaN  2021                   25.660398
22  ASEAN (Ember)  NaN  2022                   27.722073
23  ASEAN (Ember)  NaN  2023                   25.554388
24    Afghanistan  AFG  2000                   64.583336
25    Afghanistan  AFG  2001                   72.463770
26    Afghanistan  AFG  2002                   78.873245
27    Afghanistan  AFG  2003                   69.230770
28    Afghanistan  AFG  2004                   70.886070
29    Afghanistan  AFG  2005                   71.951220

In [31]:
df_elect.shape

(7152, 4)

In [32]:
df_elect.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7152 entries, 0 to 7151
Data columns (total 4 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Entity                      7152 non-null   object 
 1   Code                        5998 non-null   object 
 2   Year                        7152 non-null   int64  
 3   Renewables - % electricity  7152 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 223.6+ KB


In [33]:
df_elect.columns

Index(['Entity', 'Code', 'Year', 'Renewables - % electricity'], dtype='object')

In [34]:
df_elect.describe()

Year  Renewables - % electricity
count  7152.000000                 7152.000000
mean   2007.847735                   29.661348
std       9.572671                   30.555237
min    1985.000000                    0.000000
25%    2001.000000                    3.333333
50%    2009.000000                   18.332480
75%    2016.000000                   50.000000
max    2023.000000                  100.000010

In [35]:
df_elect.value_counts('Entity')

Entity
Poland            39
Romania           39
Europe (EI)       39
Italy             39
Thailand          39
                  ..
Martinique        22
East Timor        20
Montenegro        19
South Sudan       11
Western Sahara    10
Name: count, Length: 250, dtype: int64

In [36]:
df_elect.value_counts('Year')

Year
2013    249
2015    249
2006    249
2005    249
2009    249
2012    249
2008    249
2014    249
2007    249
2016    249
2017    249
2018    249
2019    249
2020    249
2021    249
2011    248
2010    248
2003    248
2004    248
2000    247
2001    247
2002    247
2022    245
2023    119
1994    103
1993    103
1992    103
1991    103
1990    103
1999    102
1998    102
1997    102
1996    102
1995    102
1986     59
1989     59
1988     59
1987     59
1985     59
Name: count, dtype: int64

### Comment


<hr style="border: none; height: 3px; background-color: #333;" />

## 3. Data wrangling 

### 3.1. Constraints

### Time Interval: 2010 to 2020

### Countries of interest 
    Brazil
    Argentina
    United States of America
    Mexico
    Germany
    Portugal
    Egypt
    Sudan
    China
    Japan
    Oman
    Iran
    Australia
    Papua New Guinea

In [37]:
# Creating a list containing the countries I am interested in

selected_countries =  [ 
    'Brazil',
    'Argentina',
    'United States of America', 'United States',
    'Mexico',
    'Germany',
    'Portugal',
    'Egypt',
    'Sudan',
    'China',
    'Japan',
    'Oman',
    'Iran',
    'Australia',
    'Papua New Guinea',
]

In [38]:
# Ensuring each datafram has a Country column 

df_co2.rename(columns={'country': 'Country'}, inplace=True)

df_hdi.rename(columns={'Entity': 'Country'}, inplace=True)

df_forest.rename(columns={'Entity': 'Country'}, inplace=True)

df_deforest.rename(columns={'Entity': 'Country'}, inplace=True)

df_energy.rename(columns={'Entity': 'Country'}, inplace=True)

df_elect.rename(columns={'Entity': 'Country'}, inplace=True)

In [39]:
# Ensuring each datafram has a Year column 

df_co2.rename(columns={'year': 'Year'}, inplace=True)

### 3.2. Subsetting CO2

In [40]:
# First narrowing down the time span

df_co2_wrangled = df_co2[df_co2['Year'].between(2010, 2020)]

df_co2_wrangled.shape

(2816, 79)

In [41]:
# Checking what unique countries were considered

unique_countries_co2 = df_co2_wrangled['Country'].unique()

In [42]:
# Transforming it to a list

countries_list_co2 = unique_countries_co2.tolist()

In [43]:
# Print the list to control

print(countries_list_co2)

['Afghanistan', 'Africa', 'Africa (GCP)', 'Albania', 'Algeria', 'Andorra', 'Angola', 'Anguilla', 'Antarctica', 'Antigua and Barbuda', 'Argentina', 'Armenia', 'Aruba', 'Asia', 'Asia (GCP)', 'Asia (excl. China and India)', 'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bolivia', 'Bonaire Sint Eustatius and Saba', 'Bosnia and Herzegovina', 'Botswana', 'Brazil', 'British Virgin Islands', 'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon', 'Canada', 'Cape Verde', 'Central African Republic', 'Central America (GCP)', 'Chad', 'Chile', 'China', 'Christmas Island', 'Colombia', 'Comoros', 'Congo', 'Cook Islands', 'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba', 'Curacao', 'Cyprus', 'Czechia', 'Democratic Republic of Congo', 'Denmark', 'Djibouti', 'Dominica', 'Dominican Republic', 'East Timor', 'Ecuador', 'Egypt', 'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia', 'Esw

## Massive problem - this data set also include groupings such as "Europe" or "World" or "Upper-middle-income countries. 
## I will manually exclude everything that does not appears to be a country/ island/ entity from the listing and filter again the CO2 dataset

In [44]:
only_countries =  [ 
    'Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola', 'Anguilla', 'Antigua and Barbuda', 'Argentina', 'Armenia', 
 'Aruba', 'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 
 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bolivia', 'Bonaire Sint Eustatius and Saba', 'Bosnia and Herzegovina', 
 'Botswana', 'Brazil', 'British Virgin Islands', 'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon', 
 'Canada', 'Cape Verde', 'Central African Republic', 'Chad', 'Chile', 'China', 'Christmas Island', 'Colombia', 'Comoros', 
 'Congo', 'Cook Islands', 'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba', 'Curacao', 'Cyprus', 'Czechia', 
 'Democratic Republic of Congo', 'Denmark', 'Djibouti', 'Dominica', 'Dominican Republic', 'East Timor', 'Ecuador', 
 'Egypt', 'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia', 'Eswatini', 'Ethiopia', 'Faroe Islands', 'Fiji', 
 'Finland', 'France', 'French Polynesia', 'Gabon', 'Gambia', 'Georgia', 'Germany', 'Ghana', 'Greece', 'Greenland', 
 'Grenada', 'Guatemala', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Honduras', 'Hong Kong', 'Hungary', 'Iceland', 
 'India', 'Indonesia',  'Iran', 'Iraq', 'Ireland', 'Israel', 'Italy', 'Jamaica', 'Japan', 'Jordan', 
 'Kazakhstan', 'Kenya', 'Kiribati', 'Kosovo', 'Kuwait', 'Kuwaiti Oil Fires', 'Kyrgyzstan', 'Laos', 'Latvia', 'Lebanon', 
 'Leeward Islands', 'Lesotho', 'Liberia', 'Libya', 'Liechtenstein', 'Lithuania', 'Luxembourg', 'Macao', 'Madagascar', 
 'Malawi', 'Malaysia', 'Maldives', 'Mali', 'Malta', 'Marshall Islands', 'Mauritania', 'Mauritius', 'Mexico', 
 'Micronesia (country)', 'Moldova', 'Monaco', 'Mongolia', 'Montenegro', 'Montserrat', 'Morocco', 
 'Mozambique', 'Myanmar', 'Namibia', 'Nauru', 'Nepal', 'Netherlands', 'New Caledonia', 'New Zealand', 'Nicaragua', 
 'Niger', 'Nigeria', 'Niue', 'North Korea', 'North Macedonia', 'Norway', 'Oman', 'Pakistan', 'Palau', 'Palestine', 
 'Panama', 'Panama Canal Zone', 'Papua New Guinea', 'Paraguay', 'Peru', 'Philippines', 'Poland', 'Portugal', 'Qatar', 
 'Romania', 'Russia', 'Rwanda', 'Ryukyu Islands', 'Saint Helena', 'Saint Kitts and Nevis', 'Saint Lucia', 
 'Saint Pierre and Miquelon', 'Saint Vincent and the Grenadines', 'Samoa', 'San Marino', 'Sao Tome and Principe', 
 'Saudi Arabia', 'Senegal', 'Serbia', 'Seychelles', 'Sierra Leone', 'Singapore', 'Sint Maarten (Dutch part)', 
 'Slovakia', 'Slovenia', 'Solomon Islands', 'Somalia', 'South Africa', 'South Korea', 'South Sudan', 'Spain', 
 'Sri Lanka', 'St. Kitts-Nevis-Anguilla', 'Sudan', 'Suriname', 'Sweden', 'Switzerland', 'Syria', 'Taiwan', 'Tajikistan', 
 'Tanzania', 'Thailand', 'Togo', 'Tonga', 'Trinidad and Tobago', 'Tunisia', 'Turkey', 'Turkmenistan', 
 'Turks and Caicos Islands', 'Tuvalu', 'Uganda', 'Ukraine', 'United Arab Emirates', 'United Kingdom', 'United States', 
 'Uruguay', 'Uzbekistan', 'Vanuatu', 'Vatican', 'Venezuela', 'Vietnam', 'Wallis and Futuna', 
 'Yemen', 'Zambia', 'Zimbabwe',
                    ]

In [45]:
# Here is a smaller selection of countries
df_co2_wrangled = df_co2_wrangled[df_co2_wrangled['Country'].isin(only_countries)]

df_co2_wrangled.shape

(2453, 79)

### 3.3. Subsetting df_hdi

In [46]:
# Filtering the countries

df_hdi_filtered = df_hdi[df_hdi['Country'].isin(selected_countries)]

print(df_hdi_filtered)

            Country Code  Year  Inequality-adjusted Human Development Index
65        Argentina  ARG  2010                                        0.716
66        Argentina  ARG  2011                                        0.725
67        Argentina  ARG  2012                                        0.734
68        Argentina  ARG  2013                                        0.737
69        Argentina  ARG  2014                                        0.739
...             ...  ...   ...                                          ...
1993  United States  USA  2018                                        0.815
1994  United States  USA  2019                                        0.818
1995  United States  USA  2020                                        0.808
1996  United States  USA  2021                                        0.815
1997  United States  USA  2022                                        0.823

[172 rows x 4 columns]


In [47]:
# Filtering the time span

df_hdi_wrangled = df_hdi_filtered[df_hdi_filtered['Year'].between(2010, 2020)]

df_hdi_wrangled.shape

(144, 4)

In [48]:
# Also checking here what countries were considered in the original dataset

unique_countries_hdi = df_hdi['Country'].unique()

countries_list_hdi = unique_countries_hdi.tolist()

print(countries_list_hdi) 

['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola', 'Arab States (UNDP)', 'Argentina', 'Armenia', 'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Benin', 'Bhutan', 'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil', 'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon', 'Canada', 'Cape Verde', 'Central African Republic', 'Chad', 'Chile', 'China', 'Colombia', 'Comoros', 'Congo', 'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cyprus', 'Czechia', 'Democratic Republic of Congo', 'Denmark', 'Djibouti', 'Dominican Republic', 'East Asia and the Pacific (UNDP)', 'East Timor', 'Ecuador', 'Egypt', 'El Salvador', 'Estonia', 'Eswatini', 'Ethiopia', 'Europe and Central Asia (UNDP)', 'Fiji', 'Finland', 'France', 'Gabon', 'Gambia', 'Georgia', 'Germany', 'Ghana', 'Greece', 'Guatemala', 'Guinea', 'Guinea-Bissau', 'Haiti', 'High human development (UNDP)', 'Honduras', 'Hong Kong', 'Hungary', 'Iceland', 'India', 'Indonesia', 'Ir

### 3.4. Subsetting df_forest

In [49]:
# Filtering the countries

df_forest_filtered = df_forest[df_forest['Country'].isin(selected_countries)]

print(df_forest_filtered)

            Country Code  Year  Forest cover
310       Argentina  ARG  1990     12.863715
311       Argentina  ARG  1991     12.796992
312       Argentina  ARG  1992     12.730269
313       Argentina  ARG  1993     12.663546
314       Argentina  ARG  1994     12.596823
...             ...  ...   ...           ...
7471  United States  USA  2016     33.893163
7472  United States  USA  2017     33.886604
7473  United States  USA  2018     33.880045
7474  United States  USA  2019     33.873486
7475  United States  USA  2020     33.866926

[446 rows x 4 columns]


In [50]:
# Filtering the time span

df_forest_wrangled = df_forest_filtered[df_forest_filtered['Year'].between(2010, 2020)]

df_forest_wrangled.shape

(152, 4)

### 3.4. Subsetting df_deforest

In [51]:
# Filtering the countries

df_deforest_filtered = df_deforest[df_deforest['Country'].isin(selected_countries)]

print(df_deforest_filtered)

              Country Code  Year  Deforestation
8           Argentina  ARG  1990       213600.0
9           Argentina  ARG  2000       327400.0
10          Argentina  ARG  2010       242400.0
11          Argentina  ARG  2015       134800.0
20          Australia  AUS  1990       626200.0
21          Australia  AUS  2000       632300.0
22          Australia  AUS  2010       416840.0
54             Brazil  BRA  1990      4254800.0
55             Brazil  BRA  2000      5129300.0
56             Brazil  BRA  2010      1867800.0
57             Brazil  BRA  2015      1695700.0
89              China  CHN  1990       327540.0
90              China  CHN  2000       312760.0
91              China  CHN  2010       329970.0
92              China  CHN  2015       133200.0
166           Germany  DEU  1990         6800.0
167           Germany  DEU  2000         5830.0
168           Germany  DEU  2010         7000.0
169           Germany  DEU  2015         7000.0
277            Mexico  MEX  1990       6

In [52]:
# Filtering the time span

df_deforest_wrangled = df_deforest_filtered[df_deforest_filtered['Year'].between(2010, 2020)]

df_deforest_wrangled.shape

(16, 4)

### 3.4. Subsetting df_energy

In [53]:
# Filtering the countries

df_energy_filtered = df_energy[df_energy['Country'].isin(selected_countries)]

print(df_energy_filtered)

            Country Code  Year  Renewables (% equivalent primary energy)
177       Argentina  ARG  1965                                  1.102694
178       Argentina  ARG  1966                                  1.081866
179       Argentina  ARG  1967                                  1.076531
180       Argentina  ARG  1968                                  1.227290
181       Argentina  ARG  1969                                  1.065133
...             ...  ...   ...                                       ...
4556  United States  USA  2019                                  9.312867
4557  United States  USA  2020                                 10.519768
4558  United States  USA  2021                                 10.502109
4559  United States  USA  2022                                 11.292717
4560  United States  USA  2023                                 11.655809

[624 rows x 4 columns]


In [54]:
# Filtering the time span

df_energy_wrangled = df_energy_filtered[df_energy_filtered['Year'].between(2010, 2020)]

df_energy_wrangled.shape

(121, 4)

### 3.4. Subsetting df_elect

In [55]:
# Filtering the countries

df_elect_filtered = df_elect[df_elect['Country'].isin(selected_countries)]

print(df_elect_filtered)

            Country Code  Year  Renewables - % electricity
274       Argentina  ARG  1990                   34.989185
275       Argentina  ARG  1991                   29.710293
276       Argentina  ARG  1992                   34.389725
277       Argentina  ARG  1993                   38.572450
278       Argentina  ARG  1994                   41.732450
...             ...  ...   ...                         ...
6770  United States  USA  2019                   18.294624
6771  United States  USA  2020                   20.315693
6772  United States  USA  2021                   20.742870
6773  United States  USA  2022                   22.350597
6774  United States  USA  2023                   22.672834

[484 rows x 4 columns]


In [56]:
# Filtering the time span

df_elect_wrangled = df_elect_filtered[df_elect_filtered['Year'].between(2010, 2020)]

df_elect_wrangled.shape

(154, 4)

<hr style="border: none; height: 3px; background-color: #333;" />

## 4. Cleaning

### 4.1. What I am cleaning
The cleaning steps are:
* Check for duplicates
* Check for missing values
* Check for mixed-type data

### 4.2. Cleaning df_co2_wrangled

#### Duplicated data - none

In [57]:
# Finding the duplicates
df_co2_dups = df_co2_wrangled[df_co2_wrangled.duplicated()]
df_co2_dups.shape

(0, 79)

#### Missing values - lots of them

I will acknowledge that there are a lot of missing values, but I will leave as it is and only do something about it if they disturb my analysis. I will definitely do not use all the columns and maybe also all the countries will not be necessary.

In [58]:
# Finding missing values
df_co2_missing = df_co2_wrangled.isnull().sum()

print(df_co2_missing)

Country                           0
Year                              0
iso_code                         66
population                       77
gdp                             649
                               ... 
temperature_change_from_n2o     242
total_ghg                       330
total_ghg_excluding_lucf        330
trade_co2                      1136
trade_co2_share                1136
Length: 79, dtype: int64


In [59]:
df_co2_wrangled.head(50)

Country  Year iso_code  population           gdp  cement_co2  \
160   Afghanistan  2010      AFG  28189672.0  4.739942e+10       0.015   
161   Afghanistan  2011      AFG  29249156.0  5.332634e+10       0.015   
162   Afghanistan  2012      AFG  30466484.0  5.916690e+10       0.029   
163   Afghanistan  2013      AFG  31541216.0  6.299370e+10       0.036   
164   Afghanistan  2014      AFG  32716214.0  6.434611e+10       0.029   
165   Afghanistan  2015      AFG  33753500.0  6.278339e+10       0.041   
166   Afghanistan  2016      AFG  34636212.0  6.437222e+10       0.076   
167   Afghanistan  2017      AFG  35643420.0  6.516928e+10       0.045   
168   Afghanistan  2018      AFG  36686788.0  6.599674e+10       0.057   
169   Afghanistan  2019      AFG  37769496.0  7.308571e+10       0.038   
170   Afghanistan  2020      AFG  38972236.0  7.141771e+10       0.061   
779       Albania  2010      ALB   2913402.0  2.674782e+10       0.838   
780       Albania  2011      ALB   2900660.0  2.743105e+10       0.976   
781       Albania  2012      ALB   2892191.0  2.781981e+10       1.039   
782       Albania  2013      ALB   2887019.0  2.809856e+10       1.129   
783       Albania  2014      ALB   2884100.0  2.859717e+10       1.045   
784       Albania  2015      ALB   2882482.0  2.923167e+10       0.954   
785       Albania  2016      ALB   2881064.0  3.020069e+10       0.880   
786       Albania  2017      ALB   2879361.0  3.134898e+10       1.011   
787       Albania  2018      ALB   2877019.0  3.260900e+10       0.928   
788       Albania  2019      ALB   2873883.0  3.328975e+10       0.928   
789       Albania  2020      ALB   2866850.0  3.213073e+10       1.322   
952       Algeria  2010      DZA  35856348.0  4.507058e+11       8.252   
953       Algeria  2011      DZA  36543548.0  4.820287e+11       8.187   
954       Algeria  2012      DZA  37260568.0  4.984176e+11       7.697   
955       Algeria  2013      DZA  38000628.0  5.123733e+11       7.298   
956       Algeria  2014      DZA  38760168.0  5.318435e+11       7.668   
957       Algeria  2015      DZA  39543148.0  5.515217e+11       8.062   
958       Algeria  2016      DZA  40339328.0  5.691704e+11       9.105   
959       Algeria  2017      DZA  41136548.0  5.771388e+11      10.084   
960       Algeria  2018      DZA  41927008.0  5.840644e+11       9.746   
961       Algeria  2019      DZA  42705372.0  5.899051e+11      10.100   
962       Algeria  2020      DZA  43451668.0  5.598199e+11       8.958   
1225      Andorra  2010      AND     71542.0           NaN       0.000   
1226      Andorra  2011      AND     70587.0           NaN       0.000   
1227      Andorra  2012      AND     71035.0           NaN       0.000   
1228      Andorra  2013      AND     71385.0           NaN       0.000   
1229      Andorra  2014      AND     71644.0           NaN       0.000   
1230      Andorra  2015      AND     71766.0           NaN       0.000   
1231      Andorra  2016      AND     72563.0           NaN       0.000   
1232      Andorra  2017      AND     73853.0           NaN       0.000   
1233      Andorra  2018      AND     75034.0           NaN       0.000   
1234      Andorra  2019      AND     76361.0           NaN       0.000   
1235      Andorra  2020      AND     77723.0           NaN       0.000   
1398       Angola  2010      AGO  23364196.0  1.269842e+11       0.578   
1399       Angola  2011      AGO  24259120.0  1.404773e+11       0.765   
1400       Angola  2012      AGO  25188298.0  1.524768e+11       1.147   
1401       Angola  2013      AGO  26147012.0  1.600321e+11       1.557   
1402       Angola  2014      AGO  27128340.0  1.677504e+11       1.976   
1403       Angola  2015      AGO  28127724.0  1.693340e+11       2.030   

      cement_co2_per_capita      co2  co2_growth_abs  co2_growth_prct  ...  \
160                   0.001    8.365           1.973           30.866  ...   
161                   0.000   11.838           3.474           41.525  ...   
1

#### Mixed-type columns - iso-code was corrected

In [60]:
# Checking if dataframe contains mixed-type columns

# a 'weird' variable is created to test if the data types are consistent
# if 'weird' is true, is will print the number of the mixed-type column

for col in df_co2_wrangled.columns.tolist():
  weird = (df_co2_wrangled[[col]].map(type) != df_co2_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_co2_wrangled[weird]) > 0:
    print (col)

iso_code


#### The column iso_code has mixed type data, I will check what is supposed to be and convert the column

In [61]:
df_co2_wrangled.dtypes

Country                         object
Year                             int64
iso_code                        object
population                     float64
gdp                            float64
                                ...   
temperature_change_from_n2o    float64
total_ghg                      float64
total_ghg_excluding_lucf       float64
trade_co2                      float64
trade_co2_share                float64
Length: 79, dtype: object

In [62]:
df_co2_wrangled['iso_code'] = df_co2_wrangled['iso_code'].astype(str)

In [63]:
# Checking if dataframe contains mixed-type columns

# a 'weird' variable is created to test if the data types are consistent
# if 'weird' is true, is will print the number of the mixed-type column

for col in df_co2_wrangled.columns.tolist():
  weird = (df_co2_wrangled[[col]].map(type) != df_co2_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_co2_wrangled[weird]) > 0:
    print (col)

#### No mixed data in the data set

### 4.3. Cleaning df_hdi_wrangled

#### Duplicates - none

In [64]:
# Finding the duplicates
df_hdi_dups = df_hdi_wrangled[df_hdi_wrangled.duplicated()]
df_hdi_dups.shape

(0, 4)

#### Missing values - none

In [65]:
# Finding missing values
df_hdi_missing = df_hdi_wrangled.isnull().sum()

print(df_hdi_missing)

Country                                        0
Code                                           0
Year                                           0
Inequality-adjusted Human Development Index    0
dtype: int64


#### Mixed-type columns - none

In [66]:
# Checking if dataframe contains mixed-type columns

# a 'weird' variable is created to test if the data types are consistent
# if 'weird' is true, is will print the number of the mixed-type column

for col in df_hdi_wrangled.columns.tolist():
  weird = (df_hdi_wrangled[[col]].map(type) != df_hdi_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_hdi_wrangled[weird]) > 0:
    print (col)

### 4.4. Cleaning df_forest_wrangled

#### Duplicates - none

In [67]:
# Finding the duplicates
df_forest_dups = df_forest_wrangled[df_forest_wrangled.duplicated()]
df_forest_dups.shape

(0, 4)

#### Missing values - none

In [68]:
# Finding missing values
df_forest_missing = df_forest_wrangled.isnull().sum()

print(df_forest_missing)

Country         0
Code            0
Year            0
Forest cover    0
dtype: int64


#### Mixed-type columns - none

In [69]:
# Checking if dataframe contains mixed-type columns

# a 'weird' variable is created to test if the data types are consistent
# if 'weird' is true, is will print the number of the mixed-type column

for col in df_forest_wrangled.columns.tolist():
  weird = (df_forest_wrangled[[col]].map(type) != df_forest_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_forest_wrangled[weird]) > 0:
    print (col)

### 4.5. Cleaning df_deforest_wrangled

#### Duplicates - none

In [70]:
# Finding the duplicates
df_deforest_dups = df_deforest_wrangled[df_deforest_wrangled.duplicated()]
df_deforest_dups.shape

(0, 4)

#### Missing values - none

In [71]:
# Finding missing values
df_deforest_missing = df_deforest_wrangled.isnull().sum()

print(df_deforest_missing)

Country          0
Code             0
Year             0
Deforestation    0
dtype: int64


#### Mixed-type columns - none

In [72]:
# Checking if dataframe contains mixed-type columns

# a 'weird' variable is created to test if the data types are consistent
# if 'weird' is true, is will print the number of the mixed-type column

for col in df_deforest_wrangled.columns.tolist():
  weird = (df_deforest_wrangled[[col]].map(type) != df_deforest_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_deforest_wrangled[weird]) > 0:
    print (col)

### 4.6. Cleaning df_energy_wrangled

#### Duplicates - none

In [73]:
# Finding the duplicates
df_energy_dups = df_energy_wrangled[df_energy_wrangled.duplicated()]
df_energy_dups.shape

(0, 4)

#### Missing values - none

In [74]:
# Finding missing values
df_energy_missing = df_energy_wrangled.isnull().sum()

print(df_energy_missing)

Country                                     0
Code                                        0
Year                                        0
Renewables (% equivalent primary energy)    0
dtype: int64


#### Mixed-type columns - none

In [75]:
# Checking if dataframe contains mixed-type columns

# a 'weird' variable is created to test if the data types are consistent
# if 'weird' is true, is will print the number of the mixed-type column

for col in df_deforest_wrangled.columns.tolist():
  weird = (df_deforest_wrangled[[col]].map(type) != df_deforest_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_deforest_wrangled[weird]) > 0:
    print (col)

### 4.6. Cleaning df_elect_wrangled

#### Duplicates - none

In [76]:
# Finding the duplicates
df_elect_dups = df_elect_wrangled[df_elect_wrangled.duplicated()]
df_elect_dups.shape

(0, 4)

#### Missing values - none

In [77]:
# Finding missing values
df_elect_missing = df_elect_wrangled.isnull().sum()

print(df_elect_missing)

Country                       0
Code                          0
Year                          0
Renewables - % electricity    0
dtype: int64


#### Mixed-type columns - none

In [78]:
# Checking if dataframe contains mixed-type columns

# a 'weird' variable is created to test if the data types are consistent
# if 'weird' is true, is will print the number of the mixed-type column

for col in df_elect_wrangled.columns.tolist():
  weird = (df_elect_wrangled[[col]].map(type) != df_elect_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_elect_wrangled[weird]) > 0:
    print (col)

<hr style="border: none; height: 3px; background-color: #333;" />

## 5. Data Understanding


### 5.1. What features I am checking
* info - data types
* describe - basic statistics

### 5.2. Key info df_co2

In [79]:
df_co2_wrangled.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2453 entries, 160 to 47412
Data columns (total 79 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Country                                    2453 non-null   object 
 1   Year                                       2453 non-null   int64  
 2   iso_code                                   2453 non-null   object 
 3   population                                 2376 non-null   float64
 4   gdp                                        1804 non-null   float64
 5   cement_co2                                 2298 non-null   float64
 6   cement_co2_per_capita                      2298 non-null   float64
 7   co2                                        2354 non-null   float64
 8   co2_growth_abs                             2354 non-null   float64
 9   co2_growth_prct                            2354 non-null   float64
 10  co2_including_luc         

In [80]:
df_co2_wrangled.describe()

Year    population           gdp   cement_co2  \
count  2453.000000  2.376000e+03  1.804000e+03  2298.000000   
mean   2015.000000  3.432790e+07  6.515238e+11     7.039644   
std       3.162922  1.350162e+08  2.129986e+12    52.925443   
min    2010.000000  1.833000e+03  5.444287e+08     0.000000   
25%    2012.000000  7.610282e+05  2.756526e+10     0.000000   
50%    2015.000000  6.275342e+06  9.052598e+10     0.469000   
75%    2018.000000  2.337877e+07  4.184048e+11     2.096750   
max    2020.000000  1.424930e+09  2.415184e+13   858.233000   

       cement_co2_per_capita           co2  co2_growth_abs  co2_growth_prct  \
count            2298.000000   2354.000000     2354.000000      2354.000000   
mean                0.111822    160.016285        1.559339         2.367602   
std                 0.151380    799.088814       33.816524        12.588857   
min                 0.000000      0.004000     -547.517000       -52.657000   
25%                 0.000000      1.260250       -0.290250        -3.126500   
50%                 0.064000      8.658000        0.011000         1.089500   
75%                 0.160750     55.678500        0.692750         6.814000   
max                 1.004000  10914.012000      911.782000       141.744000   

       co2_including_luc  co2_including_luc_growth_abs  ...  \
count        2134.000000                   2134.000000  ...   
mean          205.383900                      1.372382  ...   
std           877.114014                     54.329973  ...   
min            -7.961000                   -752.701000  ...   
25%             5.372250                     -2.482500  ...   
50%            26.757500                     -0.022000  ...   
75%            88.073500                      1.138250  ...   
max         11743.429000                    961.287000  ...   

       share_global_other_co2  share_of_temperature_change_from_ghg  \
count              506.000000                           2420.000000   
mean                 2.173921                              0.454601   
std                  7.960344                              1.601654   
min                  0.000000                              0.000000   
25%                  0.126250                              0.010000   
50%                  0.391000                              0.082000   
75%                  1.278750                              0.261250   
max                 58.139000                             18.869000   

       temperature_change_from_ch4  temperature_change_from_co2  \
count                  2211.000000                  2420.000000   
mean                      0.001873                     0.004572   
std                       0.005544                     0.018058   
min                      -0.001000                     0.000000   
25%                       0.000000                     0.000000   
50%                       0.000000                     0.001000   
75%                       0.001000                     0.002000   
max                       0.060000                     0.232000   

       temperature_change_from_ghg  temperature_change_from_n2o     total_ghg  \
count                  2420.000000                  2211.000000   2123.000000   
mean                      0.006651                     0.000291    236.553066   
std                       0.023513                     0.001087    950.623930   
min                       0.000000                     0.000000   -186.660000   
25%                       0.000000                     0.000000      8.950000   
50%                       0.001000                     0.000000     38.200000   
75%                       0.004000                     0.000000    124.765000   
max                       0.278000                     0.010000  12295.620000   

       total_ghg_excluding_lucf    trade_co2  trade_co2_share  
count               2123.000000  1317.000000      1317.000000  
mean                 230.689336    -0.245128        29.101863  


### 5.3. Key info df_hdi

In [81]:
df_hdi_wrangled.info()

<class 'pandas.core.frame.DataFrame'>
Index: 144 entries, 65 to 1995
Data columns (total 4 columns):
 #   Column                                       Non-Null Count  Dtype  
---  ------                                       --------------  -----  
 0   Country                                      144 non-null    object 
 1   Code                                         144 non-null    object 
 2   Year                                         144 non-null    int64  
 3   Inequality-adjusted Human Development Index  144 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 5.6+ KB


In [82]:
df_hdi_wrangled.describe()

Year  Inequality-adjusted Human Development Index
count   144.000000                                   144.000000
mean   2015.222222                                     0.662500
std       3.132158                                     0.159494
min    2010.000000                                     0.298000
25%    2013.000000                                     0.574500
50%    2015.000000                                     0.662500
75%    2018.000000                                     0.812500
max    2020.000000                                     0.879000

### 5.4. Key info df_forest

In [83]:
df_forest_wrangled.info()

<class 'pandas.core.frame.DataFrame'>
Index: 152 entries, 330 to 7475
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Country       152 non-null    object 
 1   Code          152 non-null    object 
 2   Year          152 non-null    int64  
 3   Forest cover  152 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 5.9+ KB


In [84]:
df_forest_wrangled.describe()

Year  Forest cover
count   152.000000    152.000000
mean   2015.059211     29.690373
std       3.150178     24.413098
min    2010.000000      0.008078
25%    2012.000000     10.365027
50%    2015.000000     32.658786
75%    2018.000000     36.154995
max    2020.000000     79.889779

### 5.5. Key info df_deforest

In [85]:
df_deforest_wrangled.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16 entries, 10 to 466
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Country        16 non-null     object 
 1   Code           16 non-null     object 
 2   Year           16 non-null     int64  
 3   Deforestation  16 non-null     float64
dtypes: float64(1), int64(1), object(2)
memory usage: 640.0+ bytes


In [86]:
df_deforest_wrangled.describe()

Year  Deforestation
count    16.000000   1.600000e+01
mean   2012.187500   3.487606e+05
std       2.561738   5.739312e+05
min    2010.000000   0.000000e+00
25%    2010.000000   3.287250e+04
50%    2010.000000   1.503100e+05
75%    2015.000000   2.804925e+05
max    2015.000000   1.867800e+06

### 5.6. Key info df_energy

In [87]:
df_energy_wrangled.info()

<class 'pandas.core.frame.DataFrame'>
Index: 121 entries, 222 to 4557
Data columns (total 4 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Country                                   121 non-null    object 
 1   Code                                      121 non-null    object 
 2   Year                                      121 non-null    int64  
 3   Renewables (% equivalent primary energy)  121 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 4.7+ KB


In [88]:
df_energy_wrangled.describe()

Year  Renewables (% equivalent primary energy)
count   121.000000                                121.000000
mean   2015.000000                                 12.826171
std       3.175426                                 11.715880
min    2010.000000                                  0.926311
25%    2012.000000                                  5.840176
50%    2015.000000                                  8.875714
75%    2018.000000                                 13.335284
max    2020.000000                                 48.693333

### 5.7. Key info df_elect

In [89]:
df_elect_wrangled.info()

<class 'pandas.core.frame.DataFrame'>
Index: 154 entries, 294 to 6771
Data columns (total 4 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Country                     154 non-null    object 
 1   Code                        154 non-null    object 
 2   Year                        154 non-null    int64  
 3   Renewables - % electricity  154 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 6.0+ KB


In [90]:
df_elect_wrangled.describe()

Year  Renewables - % electricity
count   154.000000                  154.000000
mean   2015.000000                   27.497394
std       3.172595                   23.383107
min    2010.000000                    0.000000
25%    2012.000000                   11.977006
50%    2015.000000                   20.085413
75%    2018.000000                   35.234106
max    2020.000000                   87.193470

<hr style="border: none; height: 3px; background-color: #333;" />

## 6. Exporting


In [91]:
# Exporting the cleaned dataframe without missing values and duplicates 

df_co2_wrangled.to_csv(os.path.join(path, '02_Data', '2-2_Prepared', 'co2_clean.csv')) 
    
df_hdi_wrangled.to_csv(os.path.join(path, '02_Data', '2-2_Prepared', 'hdi_clean.csv'))

df_forest_wrangled.to_csv(os.path.join(path, '02_Data', '2-2_Prepared', 'forest_clean.csv'))

df_deforest_wrangled.to_csv(os.path.join(path, '02_Data', '2-2_Prepared', 'deforest_clean.csv'))

df_energy_wrangled.to_csv(os.path.join(path, '02_Data', '2-2_Prepared', 'energy_clean.csv'))

df_elect_wrangled.to_csv(os.path.join(path, '02_Data', '2-2_Prepared', 'hdi_clean.csv'))